src/pages/home/ReportScreen.js

```diff
@@ -177,11 +177,20 @@ class ReportScreen extends React.Component {
                 // optimistic case
                 (prevProps.report.statusNum === CONST.REPORT.STATUS.OPEN && this.props.report.statusNum === CONST.REPORT.STATUS.CLOSED))
         ) {
-            Navigation.goBack();
-            Report.navigateToConciergeChat();
-            // isReportRemoved will prevent <FullPageNotFoundView> showing when navigating
-            this.setState({isReportRemoved: true});
-            return;
+            // We scope it to only reports that can be left.
+            // The nested "if" is intentional and aimed at improving performance.
+            const policy = prevProps.policies[`${ONYXKEYS.COLLECTION.POLICY}${prevProps.report.policyID}`];
+            const isThread = ReportUtils.isChatThread(prevProps.report);
+            const isUserCreatedPolicyRoom = ReportUtils.isUserCreatedPolicyRoom(prevProps.report);
+            const canLeaveRoom = ReportUtils.canLeaveRoom(prevProps.report, !_.isEmpty(policy));
+
+            if (isUserCreatedPolicyRoom || canLeaveRoom || isThread) {
+                Navigation.goBack();
+                Report.navigateToConciergeChat();
+                // isReportRemoved will prevent <FullPageNotFoundView> showing when navigating
+                this.setState({isReportRemoved: true});
+                return;
+            }
```

Comment:
> Why can't we replace all this with:
 if (prevOnyxReportID && !onyxReportID) {

@allroundexperts Known issue using only that check is when we split the bill and re-join the public room right after we leave it. as I described earlier, it's tricky to determine if a report is removed. when we remove a report from device A, then device B will get this response from the pusher

```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
        "value": null
      }
    ]
  }
]
```

The response will completely remove the `props.report` and make `props.reprot` equal to `defaultProps.report`

Here is the tricky part. `props.report.reportID === null` or `_.isEqual(props.report, defaultProps.report)` doesn't always mean a report is removed It can also mean:

1. User is navigating to another report
2. Onyx data for that report is not ready

Since we have a massive call of `openReport`, then the concurrent situation can happen, there is a chance that a `openReport` call already filled `props.report` but another concurrent openReport (e.g when loading report actions, or updateing lastReadReport) thinks `props.report` is not ready and then in componentDidUpdate condition `prevOnyxReportID && !onyxReportID` will be true.

In summary, since this `ReportScreen` is being used by other pages other than the room, then unexpected situations can happen.

Why scoping the condition works?
It actually minimizes the unexpected error, currently, we can't leave split bill chat, so our navigateToConcierge will not be executed.


Is there any other solution?

Yes, if we can modify the backend to send this value instead of null then we can exactly determine `isReportRemoved` condition and remove the remaining onyx data of a report from the front end. `CONST.REPORT.STATUS.REMOVED` is a new value, since `CONST.REPORT.STATUS.CLOSED` can also mean that the report is archived but not removed

```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
        value: {
                        reportID: 7794307233155968
                        stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
                        statusNum: CONST.REPORT.STATUS.REMOVED,
          },
      }
    ]
  }
]
```

Reply:
@allroundexperts Sure, here is the video

https://github.com/Expensify/App/assets/11847279/8a26bb27-f6bc-43bb-b122-46c37611b25f

1. split bill issue
2. re-join the public room right after leaving it. Notice that the issue not always occurred since `openReport` call not always executed concurrently

Reply:
@allroundexperts I'm not always reproducing 404 with split bill, what is the situation to have that 404 page?

Reply:
If we check this comment and [this issue](https://github.com/Expensify/App/issues/22223), then that hmm page might be expected. but that one is not caused by our PR

<img width="994" alt="image" src="https://github.com/Expensify/App/assets/11847279/e79bc559-3da5-4a1e-a7e1-f9a134b3f77f">

Reply:
@allroundexperts Yes, we have to distinguish it.  For leaving a report the pusher's response might be `CONST.REPORT.STATUS.REMOVED` and for discarded/canceled report due to unexpected error the state might be `CONST.REPORT.STATUS.CANCELLED`

Reply:
We have that check in our current change, but that doesn't make 2nd issue disappear

https://github.com/Expensify/App/blob/cd851d0fbd901c70a32220472104ffc887639f2a/src/pages/home/ReportScreen.js#L175

Reply:
@allroundexperts For the split bill problem, try to use high traffic account and choose three or more participants we are not having interaction with. No need to click the split bill preview, just wait right after the split bill is created, then it will redirect to the concierge unexpectedly.

Reply:
@allroundexperts Ah yeah, Sorry, for the second problem, the issue comes from report action view that sometimes concurrently triggers openReport. If we stop opening the report on componentdidmount of report action view, then we'll have an issue when [switching the report](https://github.com/Expensify/App/issues/21518#issuecomment-1633845506).

Reply:
@allroundexperts sorry for my previous comments that make you confused, I'm replying In a hurry so didn't notice which 2nd issue is. Let me clear this out.

For the 1st issue, (split bill) we need to distinguish report removal when a user leaves a room and when the report is canceled due to an unexpected error. And it will require backend change.

For the 2nd issue (re-join public group right after left), `prevOnyxReportID === routeReportID` is a check to prevent unexpected errors when navigating to another report. The 2nd issue was resolved by this check

https://github.com/Expensify/App/blob/586b647e0ade3d0f53341955e3b829447f3e807a/src/pages/home/ReportScreen.js#L175-L179

defaultProps.report is an actual null value for a component, defaultProps.report of `ReportScreen` and `ReportActionView` was different so `_.isEqual(this.props.report, defaultProps.report)` will make sure that we use the null value of `ReportScreen`. If we distinguish it from the backend, I believe we won't need this check anymore.

Reply:
@allroundexperts  When leaving a room using `LeaveRoom` API the pusher will send this response:

```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
        "value": null
      }
    ]
  }
]
```

I'd recommend that it will have this response after the change
```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
          "value": {
                        reportID: 7794307233155968 // it's required for comparison purpose
                        stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
                        statusNum: CONST.REPORT.STATUS.REMOVED, // or LEFT
          },
      }
    ]
  }
]
```
And we can add a new status

```
        STATUS: {
            OPEN: 0,
            SUBMITTED: 1,
            CLOSED: 2,
            APPROVED: 3,
            REIMBURSED: 4,
            REMOVED: 5, // or LEFT
        },
```


Optional:
If the report is removed due to cancelation or unexpected error (like split bill case) the response can be:

```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
          "value": {
                        reportID: 7794307233155968 // it's required for comparison purpose
                        stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
                        statusNum: CONST.REPORT.STATUS.CANCELLED,
          },
      }
    ]
  }
]
```

Reply:
@allroundexperts thoughts on the proposed backend change above?

Reply:
@wildan-m I'm just catching up here (haven't read all the discussions so far), would you mind explaining why we can't send the pusher response below to completely remove the report?

```
[
  {
    "eventType": "onyxApiUpdate",
    "data": [
      {
        "onyxMethod": "set",
        "key": "report_7794307233155968",
        "value": null
      }
    ]
  }
]
```

Reply:
@luacmartins the explanation to your question is [here](https://github.com/Expensify/App/pull/23051#discussion_r1271355716). Please let me know if you need more explanation on the specific parts.

Reply:
@wildan-m Thanks for the link! The proposal to add a new state won't work since we store those values in our database and they are shared across all users who have access to that report. Is there a better way to determine if the report was removed or not with a frontend only solution? Perhaps differentiating between `undefined` and `null` or something like that?

Reply:
That might be a possibility. I'd recommend we bring this conversation to Slack though so we can get more opinions from the team. 

Just a heads up that I'll be ooo until Aug 8.

Reply:
@luacmartins , @allroundexperts posted here https://expensify.slack.com/archives/C01GTK53T8Q/p1691111512068339

Reply:
@luacmartins, @allroundexperts seems there is no other suggested solution in Slack. can we proceed with [this solution](https://github.com/Expensify/App/pull/23051#discussion_r1283468361)?

Reply:
Thanks for starting the discussion in Slack @wildan-m. Let's see what others say about it.